File: Sources/AgentCoding/Mitm/HostAgentClient.swift
import Foundation

/// Tiny synchronous client for an ssh-agent unix socket. Used by
/// `SSHAgentServer` to forward SIGN_REQUESTs and REQUEST_IDENTITIES
/// from the VM into bromure's private ssh-agent — the only host-side
/// agent the VM ever sees.
///
/// The user's launchd `SSH_AUTH_SOCK` is intentionally NOT plumbed
/// through here. Earlier revisions multiplexed it in alongside the
/// private agent; that exposed every key in the user's daily-driver
/// agent (GitHub, work, prod) to the disposable VM with no consent
/// gate. Keys the user wants reachable from the VM now go through
/// the explicit-import flow in the profile UI, which lands the key
/// in `_bromurePrivate` and registers an `ImportedApproval` for
/// per-key consent.
final class HostAgentClient: @unchecked Sendable {
    let socketPath: String

    /// Set by MitmEngine after it spawns the private ssh-agent. Used
    /// as the destination for ssh-add of per-profile keys and as the
    /// only forwarding target for in-VM SIGN_REQUESTs.
    static var _bromurePrivate: HostAgentClient?

    init(socketPath: String) {
        self.socketPath = socketPath
    }

    /// Send a single ssh-agent-protocol request frame, return the
    /// response frame (without the 4-byte length prefix). nil on
    /// any I/O failure — caller falls back to its own behaviour.
    func request(_ payload: Data) -> Data? {
        let fd = socket(AF_UNIX, SOCK_STREAM, 0)
        guard fd >= 0 else { return nil }
        defer { close(fd) }

        var addr = sockaddr_un()
        addr.sun_family = sa_family_t(AF_UNIX)
        let pathBytes = Array(socketPath.utf8)
        // sockaddr_un.sun_path is 104 bytes on macOS — stay under that.
        let maxPath = MemoryLayout.size(ofValue: addr.sun_path) - 1
        guard pathBytes.count <= maxPath else { return nil }
        withUnsafeMutableBytes(of: &addr.sun_path) { dest in
            let typed = dest.bindMemory(to: CChar.self)
            for (i, b) in pathBytes.enumerated() { typed[i] = CChar(bitPattern: b) }
            typed[pathBytes.count] = 0
        }
        let connectStatus = withUnsafePointer(to: &addr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sa in
                connect(fd, sa, socklen_t(MemoryLayout<sockaddr_un>.size))
            }
        }
        if connectStatus < 0 { return nil }

        // Frame: 4-byte big-endian length + payload.
        var lenBE = UInt32(payload.count).bigEndian
        let lenData = Data(bytes: &lenBE, count: 4)
        guard writeAll(fd: fd, data: lenData),
              writeAll(fd: fd, data: payload) else { return nil }

        // Read response length, then body.
        guard let lenBuf = readExact(fd: fd, count: 4) else { return nil }
        let respLen = Int(UInt32(bigEndian: lenBuf.withUnsafeBytes { $0.load(as: UInt32.self) }))
        guard respLen > 0, respLen <= 256 * 1024 else { return nil }
        return readExact(fd: fd, count: respLen)
    }
}

private func writeAll(fd: Int32, data: Data) -> Bool {
    return data.withUnsafeBytes { (raw: UnsafeRawBufferPointer) -> Bool in
        var sent = 0
        while sent < data.count {
            let n = write(fd, raw.baseAddress!.advanced(by: sent), data.count - sent)
            if n < 0 {
                if errno == EINTR { continue }
                return false
            }
            sent += n
        }
        return true
    }
}

private func readExact(fd: Int32, count: Int) -> Data? {
    var buf = [UInt8](repeating: 0, count: count)
    var got = 0
    while got < count {
        let n = buf.withUnsafeMutableBufferPointer { ptr -> Int in
            read(fd, ptr.baseAddress!.advanced(by: got), count - got)
        }
        if n == 0 { return nil }
        if n < 0 {
            if errno == EINTR { continue }
            return nil
        }
        got += n
    }
    return Data(buf)
}

File: windows/Bromure.AC.Mitm/Ssh/HostAgentClient.cs
using System.Buffers.Binary;
using System.IO.Pipes;
using System.Net;
using System.Net.Sockets;

namespace Bromure.AC.Mitm.Ssh;

/// <summary>
/// Direct port of <c>Sources/AgentCoding/Mitm/HostAgentClient.swift</c>.
/// Tiny synchronous client for an ssh-agent endpoint. Used by
/// <see cref="SshAgentServer"/> to forward <c>SIGN_REQUEST</c> and
/// <c>REQUEST_IDENTITIES</c> from the VM into Bromure's private
/// ssh-agent — the only host-side agent the VM ever sees.
///
/// <para><b>Endpoint shape on Windows.</b> macOS uses an
/// <c>AF_UNIX</c> socket at <c>~/.bromure/agent.sock</c>. Windows
/// switches to a Named Pipe at <c>\\.\pipe\bromure-ac-ssh-agent</c>;
/// OpenSSH-for-Windows already supports named-pipe agent sockets via
/// <c>SSH_AUTH_SOCK</c> pointing at a pipe path, so guest tooling
/// doesn't notice the change.</para>
///
/// <para><b>What we don't proxy.</b> The user's regular
/// <c>SSH_AUTH_SOCK</c> is intentionally NOT plumbed through. Earlier
/// macOS revisions multiplexed it in alongside the private agent —
/// that exposed every key in the user's daily-driver agent to the
/// disposable VM with no consent gate. Keys the user wants reachable
/// from the VM go through the explicit-import flow in the profile UI.
/// </para>
/// </summary>
public sealed class HostAgentClient
{
    public string Endpoint { get; }
    public AgentEndpointKind Kind { get; }

    /// <summary>
    /// Singleton set by the engine after it spawns the private ssh-agent.
    /// Used as the destination for ssh-add of per-profile keys and as
    /// the only forwarding target for in-VM SIGN_REQUESTs.
    /// </summary>
    public static HostAgentClient? BromurePrivate { get; set; }

    public HostAgentClient(string endpoint, AgentEndpointKind kind)
    {
        Endpoint = endpoint;
        Kind = kind;
    }

    /// <summary>
    /// Send a single ssh-agent-protocol request frame; return the response
    /// frame (without the 4-byte length prefix). Null on any I/O failure
    /// — caller falls back to its own behaviour.
    /// </summary>
    public async Task<byte[]?> RequestAsync(byte[] payload, CancellationToken ct = default)
    {
        try
        {
            await using var stream = await OpenAsync(ct).ConfigureAwait(false);
            // Frame: 4-byte big-endian length + payload.
            var lenBuf = new byte[4];
            BinaryPrimitives.WriteUInt32BigEndian(lenBuf, (uint)payload.Length);
            await stream.WriteAsync(lenBuf, ct).ConfigureAwait(false);
            await stream.WriteAsync(payload, ct).ConfigureAwait(false);
            await stream.FlushAsync(ct).ConfigureAwait(false);

            var respLenBuf = await ReadExactAsync(stream, 4, ct).ConfigureAwait(false);
            if (respLenBuf is null) return null;
            var respLen = (int)BinaryPrimitives.ReadUInt32BigEndian(respLenBuf);
            if (respLen <= 0 || respLen > 256 * 1024) return null;
            return await ReadExactAsync(stream, respLen, ct).ConfigureAwait(false);
        }
        catch (IOException)
        {
            return null;
        }
        catch (SocketException)
        {
            return null;
        }
    }

    private async Task<Stream> OpenAsync(CancellationToken ct)
    {
        switch (Kind)
        {
            case AgentEndpointKind.NamedPipe:
                var pipe = new NamedPipeClientStream(".",
                    Endpoint, PipeDirection.InOut, PipeOptions.Asynchronous);
                await pipe.ConnectAsync(5000, ct).ConfigureAwait(false);
                return pipe;

            case AgentEndpointKind.UnixSocket:
                var sock = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
                await sock.ConnectAsync(new UnixDomainSocketEndPoint(Endpoint), ct).ConfigureAwait(false);
                return new NetworkStream(sock, ownsSocket: true);

            case AgentEndpointKind.LoopbackTcp:
                var (host, port) = ParseLoopback(Endpoint);
                var tcp = new TcpClient();
                await tcp.ConnectAsync(host, port, ct).ConfigureAwait(false);
                return tcp.GetStream();

            default:
                throw new InvalidOperationException("Unknown endpoint kind: " + Kind);
        }
    }

    private static (string Host, int Port) ParseLoopback(string endpoint)
    {
        // "127.0.0.1:1234"
        var colon = endpoint.LastIndexOf(':');
        if (colon < 0) throw new ArgumentException("Loopback endpoint requires host:port");
        return (endpoint[..colon], int.Parse(endpoint[(colon + 1)..], System.Globalization.CultureInfo.InvariantCulture));
    }

    private static async Task<byte[]?> ReadExactAsync(Stream stream, int count, CancellationToken ct)
    {
        var buf = new byte[count];
        var got = 0;
        while (got < count)
        {
            var n = await stream.ReadAsync(buf.AsMemory(got, count - got), ct).ConfigureAwait(false);
            if (n == 0) return null;
            got += n;
        }
        return buf;
    }
}

public enum AgentEndpointKind
{
    NamedPipe,
    UnixSocket,
    LoopbackTcp,
}
